File: src/test/java/at/searles/parsing/test/FormattingTest.java
package at.searles.parsing.test;

import at.searles.lexer.Lexer;
import at.searles.lexer.SkipTokenizer;
import at.searles.parsing.*;
import at.searles.regex.Regex;
import at.searles.regex.RegexParser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.Test;

public class FormattingTest {

    private static final Mapping<CharSequence, String> ToString = new Mapping<CharSequence, String>() {
        @Override
        public String parse(ParserStream stream, @NotNull CharSequence left) {
            return left.toString();
        }

        @Nullable
        @Override
        public CharSequence left(@NotNull String result) {
            return result;
        }
    };


    enum Annotation { BLOCK, ARGUMENT }

    @Test
    public void test() {
        // XXX this test currently only checks whether everything works without problems
        Lexer lexer = new Lexer();
        SkipTokenizer tokenizer = new SkipTokenizer(lexer);

        int ws = lexer.add(RegexParser.parse("[ \n\r\t]+"));
        tokenizer.addSkipped(ws);

        Parser<String> a = Parser.fromRegex(Regex.text("a"), tokenizer, false, ToString);
        Recognizer open = Recognizer.fromString("(", tokenizer, false);
        Recognizer close = Recognizer.fromString(")", tokenizer, false);

        Ref<String> expr = new Ref<>("expr");

        Parser<String> term = a.or(open.then(expr).annotate(Annotation.BLOCK).then(close));
        expr.set(term.then(
                Reducer.rep(
                        term
                                .annotate(Annotation.ARGUMENT)
                                .fold((stream, left, right) -> left + right)
                )
        ));

        ParserStream stream = ParserStream.fromString("a(aa((aaa)a)a)");

        stream.tokStream().setListener((src, tokId, frame) -> {

        });
    }
}

File: src/test/java/at/searles/parsing/test/FormattingTest.kt
package at.searles.parsing.test

import at.searles.buf.Frame
import at.searles.lexer.Lexer
import at.searles.lexer.SkipTokenizer
import at.searles.lexer.TokenStream
import at.searles.parsing.*
import at.searles.parsing.Parser.Companion.fromRegex
import at.searles.parsing.Reducer.Companion.rep
import at.searles.regexp.Regexp
import at.searles.regexparser.RegexpParser
import org.junit.Test

class FormattingTest {
    internal enum class Annotation {
        BLOCK, ARGUMENT
    }

    @Test
    fun test() {
        // XXX this test currently only checks whether everything works without problems
        val lexer = Lexer()
        val tokenizer = SkipTokenizer(lexer)
        val ws = lexer.add(RegexpParser.parse("[ \n\r\t]+"))
        tokenizer.addSkipped(ws)
        val a = fromRegex(Regexp.text("a"), tokenizer, false, ToString)
        val open = Recognizer.fromString("(", tokenizer, false)
        val close = Recognizer.fromString(")", tokenizer, false)
        val expr = Ref<String>("expr")
        val term = a.or(open.plus(expr).annotate(Annotation.BLOCK).plus(close))
        expr.ref = term + term.annotate(Annotation.ARGUMENT).fold(Fold.create<String, String, String> { left, right -> left + right }).rep()
        val stream: ParserStream = ParserStream.create("a(aa((aaa)a)a)")
        stream.tokStream().setListener { _: TokenStream, _: Int, _: Frame -> }
    }

    companion object {
        private val ToString: Mapping<CharSequence, String> = object : Mapping<CharSequence, String> {
            override fun parse(stream: ParserStream, input: CharSequence): String {
                return input.toString()
            }

            override fun left(result: String): CharSequence? {
                return result
            }
        }
    }
}